KAFKA-19612: Remove Redundant Validation in StripedReplicaPlacer

pjl1070048431 opened the pull request:
Currently, the following validation checks are performed twice for each partition placement:

throwInvalidReplicationFactorIfNonPositive()
throwInvalidReplicationFactorIfTooFewBrokers()
throwInvalidReplicationFactorIfZero()
These checks are already performed in the public place() method before the partition placement loop begins. Since the cluster state and replication factor don't change during the placement operation, these checks only need to be performed once.

This redundant validation could be removed from the inner loop to improve performance, especially when placing many partitions.

Delete this text and replace it with a detailed description of your change. The 
PR title and body will become the squashed commit message.

If you would like to tag individuals, add some commentary, upload images, or
include other supplemental information that should not be part of the eventual
commit message, please use a separate comment.

If applicable, please include a summary of the testing strategy (including 
rationale) for the proposed change. Unit and/or integration tests are expected
for any behavior change and system tests should be considered for larger
changes.

pjl1070048431 committed:
KAFKA-19612: Remove Redundant Validation in StripedReplicaPlacer

Currently, the following validation checks are performed twice for each partition placement:

throwInvalidReplicationFactorIfNonPositive()
throwInvalidReplicationFactorIfTooFewBrokers()
throwInvalidReplicationFactorIfZero()
These checks are already performed in the public place() method before the partition placement loop begins. Since the cluster state and replication factor don't change during the placement operation, these checks only need to be performed once.

This redundant validation could be removed from the inner loop to improve performance, especially when placing many partitions.

chia7712 reviewed (approved):
nice find. LGTM

chia7712 commented:
@pjl1070048431 Could you please merge trunk to trigger QA again?

pjl1070048431 commented:
ok

pjl1070048431 committed:
Merge remote-tracking branch 'origin/trunk' into throw_invalid_fix

DL1231 reviewed (approved):
Thanks for the patch jialun, LGTM!

m1a2st reviewed (approved):
Thanks, LGTM

JimmyWang6 reviewed (approved):
Thanks for the patch, LGTM.

chia7712 commented:
@pjl1070048431 Could you please check the CI failure?

pjl1070048431 commented:
ok i will check it

pjl1070048431 committed:
KAFKA-19612: Remove Redundant Validation in StripedReplicaPlacer

Currently, the following validation checks are performed twice for each partition placement:

throwInvalidReplicationFactorIfNonPositive()
throwInvalidReplicationFactorIfTooFewBrokers()
throwInvalidReplicationFactorIfZero()
These checks are already performed in the public place() method before the partition placement loop begins. Since the cluster state and replication factor don't change during the placement operation, these checks only need to be performed once.

This redundant validation could be removed from the inner loop to improve performance, especially when placing many partitions.

chia7712 commented:
https://github.com/apache/kafka/pull/10823/files#r646142461

@pjl1070048431 Could you please take a look at above comment? it makes sense to me since `RackList` could be access in the same package. Especially, [KIP-1203](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1203%3A+Allow+to+configure+custom+%60ReplicaPlacer%60+implementation) could encourage developers to use `RackList`

Perhaps we could add comments explaining why we need those "redundant" validation instead. WDYT?

pjl1070048431 commented:
I think either is fine. Let's keep the checks within the loop.

chia7712 commented:
> I think either is fine. Let's keep the checks within the loop.

@pjl1070048431 Do you want to close this PR or add comments explaining the duplicate checks as I mentioned?

pjl1070048431 commented:
Just close the PR
